File: doc/codes/python/examples/guide/state-check.py
import asyncio

import autd3_link_ethercrab as ethercrab
from autd3 import Client, ClientConfig
from autd3.geometry import Autd3, Geometry

# xtask:long-running  # [hide]

CHECK_INTERVAL = 0.1


async def main() -> None:
    geometry = Geometry([Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0])])

    # ANCHOR: open
    client, checker = await Client.open_with_checker(
        geometry,
        ethercrab.EtherCrabLinkOption(),
        ClientConfig(),
    )
    # ANCHOR_END: open

    try:
        # ANCHOR: poll
        last = None
        while True:
            status = await checker.check()
            if status != last:
                for i, state in enumerate(status.device_states):
                    print(f"device[{i}]: {state}")
                print(f"all operational: {status.all_op}, any lost: {status.any_lost}, recoveries: {status.recoveries}")
                last = status
            await asyncio.sleep(CHECK_INTERVAL)
        # ANCHOR_END: poll
    finally:
        await client.close()


if __name__ == "__main__":
    asyncio.run(main())

File: doc/codes/rust/examples/guide/state-check.rs
use std::time::Duration;

use autd3_rs::geometry::{Autd3, Geometry};
use autd3_rs::{Client, ClientConfig, LinkStatus};
use autd3_rs_link_ethercrab::EtherCrabLinkOption;

const CHECK_INTERVAL: Duration = Duration::from_millis(100);

#[tokio::main(flavor = "multi_thread")]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let geometry = Geometry::new(vec![Autd3::default()]);

    // ANCHOR: open
    let (client, mut checker) = Client::open_with_checker(
        &geometry,
        EtherCrabLinkOption::default(),
        ClientConfig::default(),
    )
    .await?;
    // ANCHOR_END: open

    // ANCHOR: poll
    let mut last: Option<LinkStatus> = None;
    loop {
        let status = checker.check().await?;
        if last.as_ref() != Some(&status) {
            for (i, state) in status.devices.iter().enumerate() {
                println!("device[{i}]: {state}");
            }
            println!(
                "all operational: {}, any lost: {}, recoveries: {}",
                status.all_op(),
                status.any_lost(),
                status.recoveries
            );
            last = Some(status);
        }
        tokio::time::sleep(CHECK_INTERVAL).await;
    }
    // ANCHOR_END: poll

    client.close().await?;
    Ok(())
}
